doc/ops-architecture.md

```diff
@@ -0,0 +1,56 @@
+# Ops Architecture
+
+## Top-Level Diagram
+
+![AWS Architecture Diagram](./assets/aws_arch.png)
+
+Some Remarks:
+
+- The final Spark topology (ie. How many core nodes do we need? What core nodes EC2 instance type should we use?) is eluded from this document and will be determined after the benchmark phase. This is by no means a trivial problem, but it's really hard to make an informed decision without being able to run the real payload.
+- The Sequencer machine get two network interfaces: one on the public VPC to receive the ingress, one in the private VPC to communicate with the Spark cluster.
+- The communication between the Sequencer and the Spark master node either happens through a direct TCP socket (through the private VPC), or through a Kafka topic. See the next section for more details.
+- The Spark-related EC2 machines system configuration is managed through EMR.
+- The Sequencer EC2 system configuration is managed through a NixOS closure. The closure contains the sequencer program, launched in a sandboxed systemd unit. The closure is pushed to the remote machine through a small terraform custom provider.
+- The whole infra is provisioned and configured through terraform.
+
+## Sequencer / Spark Cluster Interaction
```

Comment:
@ycryptx: What do you think about that? I need your input.

I think we have three choices here:

1. We go the bare socket route.
2. We go the Kafka route.
3. We start with the bare socket, test things out. If we realize it's too brittle, we move to Kafka.

Overall, on the Sequencer side, we'll have to manage this sync <=> async interface regardless of the transport we'll chose. We'll need the LRU map (or something equivalent) in any case.

Reply:
Thanks for this amazing write up, and apologies for replying late. I hit a snag when I realized it will be pretty much impossible to use Spark for our purposes and have been researching how we can get around it:

The issue is that the lib we use to do ZK proving can only run in a Javascript runtime. Spark only supports step operations written in Python/Java. Our easiest way forward is to use Hadoop Streaming, which allows us to specify a custom executable to do the step operations. This will allow us to run Node.js as part of the map-reduce operation (https://aws.amazon.com/blogs/big-data/node-js-streaming-mapreduce-with-amazon-emr/). 

Note that we will have to define and upload to S3 a custom EMR Bootstrap script which will download the necessary dependencies our step functions require (https://docs.aws.amazon.com/emr/latest/ManagementGuide/emr-plan-bootstrap.html)

In regards to communication between Sequencer and EMR:

You bring up great points about synchronicity and fault tolerance. The more I think about it I think for the sake of simplicity it would be ok to just upload the map-reduce input data to S3 rather than stream it to EMR. The latency imposed would probably not be consequential. Especially if the bucket is in the same region as the sequencer. If we see that it's an issue we can work to solve it later. 

Also, to clarify, the sequencer gives the user confirmation before transaction is sent to the map-reduce stage, so we don't have to worry about synchronicity there.